addons/controllers/cpi/vspherecpiconfig_utils.go

```diff
@@ -20,13 +20,58 @@ import (
 	capvv1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/v1beta1"
 	capvvmwarev1beta1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
 	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
 	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
 	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
 	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/addonconfigs/cpi/v1alpha1"
 )
 
+// ClustersToVSphereCPIConfig returns a list of Requests with VSphereCPIConfig ObjectKey based on Cluster events
+func (r *VSphereCPIConfigReconciler) ClustersToVSphereCPIConfig(o client.Object) []ctrl.Request {
+	r.Log.V(4).Info("Clusters to VSphereCPIConfig handler")
+
+	cluster, ok := o.(*clusterapiv1beta1.Cluster)
+	if !ok {
+		r.Log.Error(errors.New("invalid type"),
+			"Expected to receive Cluster resource",
+			"actualType", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	r.Log.V(4).Info("Mapping Cluster to VSphereCPIConfig")
+
+	cs := &cpiv1alpha1.VSphereCPIConfigList{}
+	_ = r.List(context.Background(), cs)
+
+	requests := []ctrl.Request{}
+	for i := 0; i < len(cs.Items); i++ {
+		config := &cs.Items[i]
+		if config.Namespace == cluster.Namespace {
+			// corresponding vsphereCPIConfig should have following ownerRef
+			ownerReference := metav1.OwnerReference{
+				APIVersion: clusterapiv1beta1.GroupVersion.String(),
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			}
+			if clusterapiutil.HasOwnerRef(config.OwnerReferences, ownerReference) || config.Name == cluster.Name {
```

Comment:
the `||` statement intended? feel a little bit hacky? also namespace should be checked as well

Reply:
The namespace is already checked here https://github.com/vmware-tanzu/tanzu-framework/pull/3318/files/e373dee7e53e0efb6ef13ed4c3d53b2d85415791#diff-dee6d9aaf480ac1ee68633bf4b312d5e2165eb7b85cda8f67255bdaea06f3559R53
We should rely on the owner reference first. Then check the name

Reply:
Should we expect Cluster.Name == Config.Name? In which case they does not equal?

Reply:
The logic first list all the Configs, so there would be configs created for other clusters

Reply:
Why don't we just call r.get() with specified namespace and name?

Reply:
Because, the VSphereCPIConfig doesn't have to be the same name as the Cluster

Reply:
https://github.com/vmware-tanzu/tanzu-framework/tree/main/addons#pre-create-clusterbootstrap-resource
Example here